kube/src/client/discovery.rs

```diff
@@ -0,0 +1,341 @@
+    /// If the group is not served or does not support this version,
+    /// returns empty vector.
+    pub fn resources_by_group_and_version(&self, g: &str, ver: &str) -> Vec<GroupVersionKind> {
+        match self.groups.get(g) {
+            Some(group) => group.resources_by_version(ver),
+            None => Vec::new(),
+        }
+    }
+
+    /// Returns resource with given group, version and kind.
+    ///
+    /// This function returns `GroupVersionKind` which can be used together
+    /// with `DynamicObject` and raw `APIResource` value with additional information.
+    pub fn resolve_group_version_kind(
+        &self,
+        group: &str,
+        version: &str,
+        kind: &str,
+    ) -> Option<(GroupVersionKind, APIResource)> {
+        // TODO: could be better than O(N)
+        let group = self.group(group)?;
+        group
+            .resources_by_version(version)
+            .into_iter()
+            .find(|gvk| gvk.kind == kind)
+            .map(|gvk| {
+                let data = group
+                    .versions_and_resources
+                    .iter()
+                    .find(|data| data.version == version)
+                    .unwrap();
+                let raw = data.list.resources.iter().find(|r| r.kind == kind).unwrap();
+                (gvk, raw.clone())
+            })
+    }
+}
+
+impl Group {
+    /// Core group name
+    pub const CORE_GROUP: &'static str = "core";
+
+    /// Returns the name of this group.
+    /// For core group (served at `/api`), returns "core" (also declared as
+    /// `Group::CORE`).
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Actually sets up order promised by `version`
+    fn sort_versions(&mut self) {
+        self.versions_and_resources
+            .sort_by_cached_key(|ver_data| Version::parse(ver_data.version.as_str()))
+    }
+
+    /// Returns served versions (e.g. `["v1", "v2beta1"]`) of this group.
+    /// This list is always non-empty, and sorted in the following order:
+    /// - Stable versions (with the last being the first)
+    /// - Beta versions (with the last being the first)
+    /// - Alpha versions (with the last being the first)
+    /// - Other versions, alphabetically
+    // Order is documented here:
+    // https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#specify-multiple-versions
+    pub fn versions(&self) -> impl Iterator<Item = &str> {
+        let versions = self.versions_and_resources.as_slice();
+        versions.iter().map(|ver_data| ver_data.version.as_str())
+    }
+
+    /// Returns preferred version for working with given group.
+    /// This function tries to use `preferredVersion` field at first,
+    /// and then uses some fallbacks if that is missing.
+    pub fn preferred_version(&self) -> &str {
+        match &self.preferred_version {
+            Some(v) => v,
+            None => self.versions().next().unwrap(),
```

Comment:
hm, a bit uncertain about standardising this unwrap. the fallback felt like a bit of a hack in the dynamic_api example, but can't remember if if was truly necessary.

Reply:
I'll expose the non-guessing variant.